deeppavlov/core/commands/train.py

```diff
@@ -278,237 +279,253 @@ def _train_batches(model: NNModel, iterator: DataLearningIterator, train_config:
-                        } for x_item, y_predicted_item, y_true_item in zip(x, y_predicted, y_true)]
-                    except NameError:
-                        log.warning('Could not log examples')
-
-                if losses:
-                    report['loss'] = sum(losses)/len(losses)
-                    losses = []
-
-                if train_config['tensorboard_log_dir'] is not None:
-                    for name, score in metrics:
-                        metric_sum = tf.Summary(value=[tf.Summary.Value(tag='every_n_epochs/' + name,
-                                                                        simple_value=score), ])
-                        tb_train_writer.add_summary(metric_sum, epochs)
-
-                    if 'loss' in report:
-                        loss_sum = tf.Summary(value=[tf.Summary.Value(tag='every_n_epochs/' + 'loss',
-                                                                        simple_value=report['loss']), ])
-                        tb_train_writer.add_summary(loss_sum, epochs)
-
-                model.process_event(event_name='after_train_log', data=report)
-                report = {'train': report}
-                print(json.dumps(report, ensure_ascii=False))
-                train_y_true.clear()
-                train_y_predicted.clear()
-
-            if train_config['val_every_n_epochs'] > 0 and epochs % train_config['val_every_n_epochs'] == 0:
-                report = _test_model(model, metrics_functions, iterator,
-                                     train_config['batch_size'], 'valid', start_time, train_config['show_examples'])
-                report['epochs_done'] = epochs
-                report['batches_seen'] = i
-                report['train_examples_seen'] = examples
-
-                metrics = list(report['metrics'].items())
-
-                if train_config['tensorboard_log_dir'] is not None:
-                    for name, score in metrics:
-                        metric_sum = tf.Summary(value=[tf.Summary.Value(tag='every_n_epochs/' + name,
-                                                                        simple_value=score), ])
-                        tb_valid_writer.add_summary(metric_sum, epochs)
-
-                m_name, score = metrics[0]
-                if improved(score, best):
-                    patience = 0
-                    log.info('New best {} of {}'.format(m_name, score))
-                    best = score
-                    log.info('Saving model')
-                    model.save()
-                    saved = True
-                else:
-                    patience += 1
-                    log.info('Did not improve on the {} of {}'.format(m_name, best))
 
-                report['impatience'] = patience
-                if train_config['validation_patience'] > 0:
-                    report['patience_limit'] = train_config['validation_patience']
+def _init_train_monitoring(train_config, metrics_functions):
```

Comment:
This could be a class with appropriate methods.

Reply:
Could you give more detail on this idea? Now I can only see that 'improved' key can be replaced with 'improved' method.

Reply:
There can also be clear_after_log method.

Reply:
train_monitoring is a class now